src/mudlet.cpp

```diff
@@ -1072,6 +1072,11 @@ void mudlet::slot_close_profile_requested(int tab)
         mpIrcClientMap[pH]->deleteLater();
     }
 
+    // Wait for disconnection to complete
+    while (pH->mTelnet.getConnectionState() != QAbstractSocket::UnconnectedState) {
+        QApplication::processEvents();
```

Comment:
This has a chance to hang the client if the disconnection doesn't work for some reason, right?

Reply:
It *should* always be possible to get to a disconnected state though shouldn't it? From the Qt documentation (for `QAbstractSocket`):

> At any time, `QAbstractSocket` has a state (returned by `state()`). The initial state is `UnconnectedState`. After calling `connectToHost()`, the socket first enters `HostLookupState`. If the host is found, `QAbstractSocket` enters `ConnectingState` and emits the `hostFound()` signal. When the connection has been established, it enters `ConnectedState` and emits `connected()`. If an error occurs at any stage, `error()` is emitted. Whenever the state changes, `stateChanged()` is emitted. For convenience, `isValid()` returns true if the socket is ready for reading and writing, but note that the socket's state must be `ConnectedState` before reading and writing can occur.

...

>To close the socket, call `disconnectFromHost()`. `QAbstractSocket` enters `QAbstractSocket::ClosingState`. After all pending data has been written to the socket, `QAbstractSocket` actually closes the socket, enters `QAbstractSocket::UnconnectedState`, and emits `disconnected()`. If you want to abort a connection immediately, discarding all pending data, call `abort()` instead. If the remote host closes the connection, `QAbstractSocket` will emit error(`QAbstractSocket::RemoteHostClosedError`), during which the socket state will still be `ConnectedState`, and then the `disconnected()` signal will be emitted.

The waits for disconnection that I have inserted are ALWAYS after we have called `cTelnet::disconnectIt()` - although for Windows when `Host::mSsslTsl` is set `cTelnet::abortConnection()` is being used instead when closing just the selected tab when multi-playing? Which is when this problem manifests - humm...! :worried: 

NB The profile closing code is ~~a mess~~ quite convoluted - there are three different paths that *are* used to close profiles down:
+ `mudlet::closeEvent(...)`
+ `mudlet::slot_close_profile_requested(...)`
+ `mudlet::force_close()`

There is additionally one more not currently used (though it *would* be for a missing `close` [profile] menu item on the `game` main menu. or a main toolbar `close` button -  
![mudlet_door_48px](https://user-images.githubusercontent.com/6163092/53656680-e8f38d00-3c4b-11e9-9aec-78e2df84193b.png)):
+ `mudlet::slot_close_profile()`

Reply:
I don't think we need even more toolbar buttons. The X on the profile tab is quite enough, it works grey for our users!

Reply:
*Only if they are multi-playing! If I want to stop playing one character on WoTMUD I have to close the whole application before I connect with a second profile - there is no provision* currently *to close a* single *profile yet keep the Mudlet application running (and drop back to the state that you'd get by pressing cancel on the connection profiles dialogue when it appears initially)...*

Reply:
That is a no-no on WoTMUD - there is a 5/15 minute rule between changing characters on the same/different sides - and breaking that can result in zapping of the characters concerned. Which OTOH does give the chance to grab a bite to eat mean-times and it wouldn't be totally the end of the World to close the application down - but it isn't the only way to do things...

Reply:
Alt+D will disconnect your current profile. It is not the same, though.